dartsim/src/Base.hh

```diff
@@ -398,29 +423,68 @@ class Base : public Implements3d<FeatureList<Feature>>
     if (_link->inertial)
     {
       std::size_t nodeCount = 1 + _link->weldedNodes.size();
-      const double mass = _link->inertial->MassMatrix().Mass() / nodeCount;
-      const math::Matrix3d moi = _link->inertial->Moi() * (1 / nodeCount);
-      const math::Vector3d cog = _link->inertial->Pose().Pos();
-
-      _link->link->setMass(mass);
-      _link->link->setMomentOfInertia(
-          moi(0, 0), moi(1, 1), moi(2, 2),
-          moi(0, 1), moi(0, 2), moi(1, 2));
-      _link->link->setLocalCOM(math::eigen3::convert(cog));
-      for (auto weldedNodePair : _link->weldedNodes)
+      const auto dividedInertial = DivideInertial(*_link->inertial, nodeCount);
+      AssignInertialToBody(dividedInertial, _link->link);
+      for (const auto &weldedNodePair : _link->weldedNodes)
       {
-        weldedNodePair.first->setMass(mass);
-        weldedNodePair.first->setMomentOfInertia(
-            moi(0, 0), moi(1, 1), moi(2, 2),
-            moi(0, 1), moi(0, 2), moi(1, 2));
-        weldedNodePair.first->setLocalCOM(math::eigen3::convert(cog));
+        AssignInertialToBody(dividedInertial, weldedNodePair.first);
       }
     }
 
     this->linkByWeldedNode[pairJointBodyNode.second] = _link;
     return pairJointBodyNode.second;
   }
 
+  public: void MergeLinkAndWeldedBody(LinkInfo *_link, DartBodyNode *child)
+  {
+    // Break the existing joint first.
+    child->moveTo<dart::dynamics::FreeJoint>(nullptr);
+    auto it = _link->weldedNodes.begin();
+    bool foundWeld = false;
+    for (; it != _link->weldedNodes.end(); ++it)
+    {
+      if (it->first == child)
+      {
+        auto worldId = this->GetWorldOfModelImpl(
+            this->models.objectToID[child->getSkeleton()]);
+        auto dartWorld = this->worlds.at(worldId);
+        dartWorld->getConstraintSolver()->removeConstraint(it->second);
+        // Okay to erase since we break afterward.
+        _link->weldedNodes.erase(it);
+        foundWeld = true;
+        break;
+      }
+    }
+
+    if (!foundWeld)
+    {
+      // We have not found a welded node associated with _link. This shouldn't
+      // happen.
+      ignerr << "Could not find welded body node for link " << _link->name
+             << ". Merging of link and welded body failed.";
+      return;
+    }
+
+    if (_link->inertial)
+    {
+      std::size_t nodeCount = _link->weldedNodes.size();
+      if (nodeCount > 0)
```

Comment:
`nodeCount` is `1 + _link->weldedNodes.size()` in `SplitAndWeldLink` and I think it should be the same here

then this if statement wouldn't be necessary

Reply:
Thanks for catching that. https://github.com/gazebosim/gz-physics/pull/367/commits/e772f869fa11ee06456dc7e5093b10e1827c2f7e